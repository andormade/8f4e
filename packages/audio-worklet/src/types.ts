export interface MidiModuleAddresses {
	moduleId: string;
	noteAddress: number | undefined;
	channelAddress: number | undefined;
	noteOnOffAddress: number | undefined;
	velocityAddress: number | undefined;
}

export interface MidiCCModuleAddresses {
	moduleId: string;
	valueAddress: number | undefined;
	channelAddress: number | undefined;
	selectedCCAddress: number | undefined;
}

export interface RNBOModule {
	patcherId: string;
	paramAdresses: number[];
}
