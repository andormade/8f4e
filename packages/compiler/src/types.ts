export interface CompiledModule {
	functionBody: number[];
	initialMemory: number[];
	memoryAddresses: { address: number; id: string; default?: number }[];
	moduleId: string;
	offset: number;
}

export type MemoryBuffer = Int32Array;

export interface Connection {
	fromModuleId: string;
	fromConnectorId: string;
	toModuleId: string;
	toConnectorId: string;
}

export type RelativeAddressCalculator = (nthWord: number) => number;

export type ModuleGenerator<TConfig = unknown> = (
	moduleId: string,
	offset: RelativeAddressCalculator,
	initialConfig?: TConfig
) => CompiledModule;

export type EngineConfig = Record<string, number | string | EngineConfig[]>;

export interface Engine {
	name: string;
	config: EngineConfig;
}

export type ModuleState = Record<string, any>;

export type ModuleStateExtractor<T> = (memoryBuffer: MemoryBuffer, moduleAddress: number) => T;
export type ModuleStateInserter<T> = (moduleState: T, memoryBuffer: MemoryBuffer, moduleAddress: number) => void;

export interface Module {
	id: string;
	engine: Engine;
	state: ModuleState;
}

export type MemoryAddressLookup = Record<string, number>;
