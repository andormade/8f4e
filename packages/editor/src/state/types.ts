import { CompileOptions, CompiledModuleLookup, MemoryBuffer, DataStructure } from '@8f4e/compiler';
import { Font, SpriteLookups } from '@8f4e/sprite-generator';
import { SpriteLookup } from '@8f4e/2d-engine';

export interface CodeBlock {
	code: string[];
	isOpen: boolean;
	x: number;
	y: number;
	viewport?: Viewport;
	codeBlocks?: CodeBlock[];
}

export interface Size {
	width: number;
	height: number;
}

export interface Position {
	x: number;
	y: number;
}

export type Viewport = {
	x: number;
	y: number;
};

interface ContextMenuButton {
	action?: string;
	close?: boolean;
	payload?: Record<string, unknown>;
	disabled?: boolean;
	divider?: boolean;
	title?: string;
	isSectionTitle?: boolean;
}

interface MenuItemDivider extends ContextMenuButton {
	divider: true;
	title: never;
}

export type ExtendedInstructionSet = 'debug' | 'button' | 'switch' | 'offset' | 'plot' | 'piano';

export type ContextMenuItem = ContextMenuButton | MenuItemDivider;

export type MenuGenerator = (state: State, payload?: any) => ContextMenuItem[] | Promise<ContextMenuItem[]>;

export interface ContextMenu extends Position {
	highlightedItem: number;
	itemWidth: number;
	items: ContextMenuItem[];
	open: boolean;
	menuStack: string[];
}

export interface BuildError {
	lineNumber: number;
	code: number;
	message: string;
	moduleId: string;
}

export interface Runtime {
	runner: 'audioWorklet' | 'webWorker';
	latency: number;
	renderQuantum: number;
}

export interface Compiler {
	codeBuffer: Uint8Array;
	compilationTime: number;
	cycleTime: number;
	isCompiling: boolean;
	lastCompilationStart: number;
	memoryBuffer: MemoryBuffer;
	memoryBufferFloat: Float32Array;
	memoryRef: WebAssembly.Memory;
	timerAccuracy: number;
	compiledModules: CompiledModuleLookup;
	buildErrors: BuildError[];
	compilerOptions: CompileOptions;
	allocatedMemorySize: number;
}

export interface Midi {
	outputs: MIDIOutput[];
	inputs: MIDIInput[];
}

export interface MemoryIdentifier {
	memory: DataStructure;
	showAddress: boolean;
	showEndAddress: boolean;
	bufferPointer: number;
	showBinary: boolean;
}

export interface BufferPlotter {
	width: number;
	height: number;
	x: number;
	y: number;
	minValue: number;
	maxValue: number;
	buffer: MemoryIdentifier;
	bufferLength: MemoryIdentifier | undefined;
}
export interface Switch {
	width: number;
	height: number;
	x: number;
	y: number;
	id: string;
	offValue: number;
	onValue: number;
}

export interface Debugger {
	width: number;
	height: number;
	showAddress: boolean;
	showEndAddress: boolean;
	x: number;
	y: number;
	id: string;
	memory: DataStructure;
	bufferPointer: number;
	showBinary: boolean;
}

export interface Output {
	codeBlock: CodeBlockGraphicData;
	width: number;
	height: number;
	x: number;
	y: number;
	id: string;
	calibratedMax: number;
	calibratedMin: number;
	memory: DataStructure;
}

export interface Input {
	codeBlock: CodeBlockGraphicData;
	width: number;
	height: number;
	x: number;
	y: number;
	id: string;
	wordAlignedAddress: number;
}

export interface PianoKeyboard {
	x: number;
	y: number;
	width: number;
	height: number;
	keyWidth: number;
	pressedKeys: Set<number>;
	pressedKeysListMemory: DataStructure;
	pressedNumberOfKeysMemory: DataStructure;
	startingNumber: number;
}

export interface CodeBlockGraphicData {
	width: number;
	minGridWidth: number;
	height: number;
	code: string[];
	trimmedCode: string[];
	padLength: number;
	codeToRender: number[][];
	codeColors: Array<Array<SpriteLookup | undefined>>;
	gaps: Map<number, { size: number }>;
	cursor: { col: number; row: number; x: number; y: number };
	id: string;
	positionOffsetterXWordAddress?: number;
	positionOffsetterYWordAddress?: number;
	x: number;
	y: number;
	offsetX: number;
	offsetY: number;
	gridX: number;
	gridY: number;
	isOpen: boolean;
	extras: {
		inputs: Map<string, Input>;
		outputs: Map<string, Output>;
		debuggers: Map<string, Debugger>;
		bufferPlotters: Map<string, BufferPlotter>;
		switches: Map<string, Switch>;
		buttons: Map<string, Switch>;
		pianoKeyboards: Map<number, PianoKeyboard>;
		errorMessages: Map<
			number,
			{
				message: string[];
				x: number;
				y: number;
			}
		>;
	};
	viewport: Viewport;
	parent: CodeBlockGraphicData;
	codeBlocks: Set<CodeBlockGraphicData>;
}

export type GraphicHelper = {
	spriteLookups?: SpriteLookups;
	outputsByWordAddress: Map<number, Output>;
	globalViewport: {
		width: number;
		height: number;
		roundedWidth: number;
		roundedHeight: number;
		vGrid: number;
		hGrid: number;
		borderLineCoordinates: {
			top: { startX: number; startY: number; endX: number; endY: number };
			right: { startX: number; startY: number; endX: number; endY: number };
			bottom: { startX: number; startY: number; endX: number; endY: number };
			left: { startX: number; startY: number; endX: number; endY: number };
		};
		center: { x: number; y: number };
	};
	baseCodeBlock: CodeBlockGraphicData;
	activeViewport: CodeBlockGraphicData;
	contextMenu: ContextMenu;
	draggedCodeBlock?: CodeBlockGraphicData;
	selectedCodeBlock?: CodeBlockGraphicData;
	dialog: {
		show: boolean;
		text: string;
		title: string;
		buttons: Array<{
			title: string;
			action: string;
			payload?: any;
		}>;
	};
};

export interface BinaryAsset {
	/** The id of the module that the binary data should be loaded into */
	moduleId?: string;
	/** The id of the memory that the binary data should be loaded into */
	memoryId?: string;
	/** The binary data in base64 format */
	data: string;
	/** The file name of the binary data */
	fileName: string;
}

export interface Project {
	title: string;
	author: string;
	description: string;
	codeBlocks: CodeBlock[];
	viewport: Viewport;
	sampleRate: number;
	binaryAssets: BinaryAsset[];
	audioInputBuffers: {
		moduleId: string;
		memoryId: string;
		channel: number;
		input: number;
	}[];
	audioOutputBuffers: {
		moduleId: string;
		memoryId: string;
		channel: number;
		output: number;
	}[];
}

export interface ExampleModule {
	title: string;
	author: string;
	code: string;
	tests: unknown[];
	category: string;
}

export interface Options {
	isLocalStorageEnabled: boolean;
	isDebugMode: boolean;
	localStorageId: string;
	exampleProjects: Record<string, Project>;
	exampleModules: Record<string, ExampleModule>;
}

export interface EditorSettings {
	colorScheme: string;
	font: Font;
}

export interface State {
	compiler: Compiler;
	runtime: Runtime;
	midi: Midi;
	graphicHelper: GraphicHelper;
	project: Project;
	options: Options;
	editorSettings: EditorSettings;
	compilationTime: number;
}
