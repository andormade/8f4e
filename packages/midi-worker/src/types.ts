export interface MidiModuleAddresses {
	moduleId: string;
	noteWordAddress: number | undefined;
	channelWordAddress: number | undefined;
	noteOnOffWordAddress: number | undefined;
	velocityWordAddress: number | undefined;
}

export interface MidiCCModuleAddresses {
	moduleId: string;
	valueWordAddress: number | undefined;
	channelWordAddress: number | undefined;
	selectedCCWordAddress: number | undefined;
}

export interface RNBOModule {
	patcherId: string;
	paramAdresses: number[];
}
