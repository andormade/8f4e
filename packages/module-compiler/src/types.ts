export const enum ArgumentType {
	LITERAL = 'literal',
	IDENTIFIER = 'identifier',
}

export type Argument = { type: ArgumentType.LITERAL; value: number } | { type: ArgumentType.IDENTIFIER; value: string };

export type AST = Array<{ instruction: string; arguments: Array<Argument> }>;

export enum MemoryTypes {
	DYNAMIC_ARRAY,
	STATIC_ARRAY,
	ARRAY_SIZE,
	INPUT_POINTER,
	OUTPUT,
	PRIVATE,
	NUMBER,
	ARRAY,
}

export type MemoryMap = Array<{
	type: MemoryTypes;
	address: number;
	byteAddress: number;
	id: string;
	size: number;
	default: number | number[];
	usage: number;
}>;
