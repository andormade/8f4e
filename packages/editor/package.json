{
	"name": "@8f4e/editor",
	"version": "1.0.0",
	"description": "8f4e",
	"author": "Andor Polgar",
	"license": "MIT",
    "scripts": {
        "build": ""
    },
	"bugs": {
		"url": "https://github.com/andormade/8f4e/issues"
	},
	"homepage": "https://github.com/andormade/8f4e#readme",
	"dependencies": {
		"@8f4e/2d-engine": "file:packages/2d-engine",
		"@8f4e/compiler": "file:packages/compiler",
		"@8f4e/sprite-generator": "file:packages/sprite-generator",
		"@8f4e/compiler-worker": "file:packages/compiler-worker",
		"@8f4e/web-worker-midi-runtime": "file:packages/web-worker-midi-runtime",
		"@8f4e/audio-worklet-runtime": "file:packages/audio-worklet-runtime",
		"buffer": "5.7.1"
	}
}
