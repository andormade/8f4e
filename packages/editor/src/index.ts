import initEvents from './events';
import humanInterface from './events/humanInterface';
import { Options, Project } from './state/types';
import initView from './view';
import initState from './state';

export default async function init(canvas: HTMLCanvasElement, project: Project, options: Partial<Options>) {
	const events = initEvents();
	humanInterface(canvas, events);
	const state = initState(events, project, options);
	const view = await initView(state, canvas);

	events.on('setColorScheme', () => {
		view.reloadSpriteSheet();
		events.dispatch('spriteSheetRerendered');
	});

	events.on('setFont', () => {
		view.reloadSpriteSheet();
		events.dispatch('spriteSheetRerendered');
	});

	return {
		resize: (width: number, height: number) => {
			events.dispatch('resize', { canvasWidth: width, canvasHeight: height });
			view.resize(width, height);
		},
		state,
	};
}
