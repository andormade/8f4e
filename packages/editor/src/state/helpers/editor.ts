import { CodeBlockGraphicData } from '../types';

export function moveCaret(
	code: string[],
	row: number,
	col: number,
	direction: 'ArrowUp' | 'ArrowDown' | 'ArrowLeft' | 'ArrowRight' | 'Jump'
): [row: number, col: number] {
	switch (direction) {
		case 'ArrowUp':
			row = Math.max(row - 1, 0);

			if (!code[row][col]) {
				col = code[row].length;
			}

			return [row, col];
		case 'ArrowDown':
			row = Math.min(row + 1, code.length - 1);

			if (!code[row][col]) {
				col = code[row].length;
			}

			return [row, col];
		case 'ArrowLeft':
			col = Math.max(col - 1, 0);
			return [row, col];
		case 'ArrowRight':
			col = Math.min(col + 1, code[row].length);
			return [row, col];
		case 'Jump':
			col = Math.max(Math.min(col, code[row].length), 0);

			if (row > code.length - 1) {
				return [code.length - 1, col];
			}

			row = row = Math.min(row, code.length - 1);
			return [row, col];
	}
}

export function backSpace(code: string[], row: number, col: number): { code: string[]; row: number; col: number } {
	const newCode = [...code];

	if (col > 0) {
		const [newRow, newCol] = moveCaret(code, row, col, 'ArrowLeft');
		newCode[row] = newCode[row].slice(0, newCol) + newCode[row].slice(newCol + 1);
		return { code: newCode, row: newRow, col: newCol };
	}

	if (row > 0) {
		const save = newCode[row];
		const [newRow] = moveCaret(newCode, row, col, 'ArrowUp');
		newCode.splice(row, 1);
		newCode[newRow] = newCode[newRow] + save;
		return { code: newCode, row: newRow, col: code[newRow].length };
	}

	return { code, row, col };
}

export function type(
	code: string[],
	row: number,
	col: number,
	char: string
): { code: string[]; row: number; col: number } {
	const newCode = [...code];
	newCode[row] = newCode[row].substring(0, col) + char + newCode[row].substring(col);
	const [newRow, newCol] = moveCaret(newCode, row, col, 'ArrowRight');
	return {
		code: newCode,
		row: newRow,
		col: newCol,
	};
}

export function enter(
	code: string[],
	row: number,
	col: number
): {
	code: string[];
	row: number;
	col: number;
} {
	const newCode = [...code];
	const save = newCode[row].substring(col);
	newCode[row] = newCode[row].substring(0, col);
	newCode.splice(row + 1, 0, '');
	newCode[row + 1] = save;

	const [newRow] = moveCaret(newCode, row, col, 'ArrowDown');
	return { code: newCode, row: newRow, col: 0 };
}

/**
 *
 *  +------------
 *  | 0 Line 0
 *  +------------
 *  | 1 Line 1   gap.size = 2
 *  +------------
 *  | 2 gap
 *  +-----------
 *  | 3 gap
 *  +-------------
 *  | 4 Line 2
 *  +--------------
 *  | 5 Line 3    gap.size = 3
 *  +--------------
 *  | 6 gap
 *  +---------------
 *  | 7 gap
 *  +------------
 *  | 8 gap
 *  +---------
 *  | 9 Line 4
 *  +----------
 *  | 10 Line 5
 *  +-------------
 * @returns
 */

export function gapCalculator(row: number, gaps: CodeBlockGraphicData['gaps']) {
	let physicalRowCounter = row;
	for (const [gapStartLine, { size }] of gaps) {
		if (row > gapStartLine) {
			physicalRowCounter += size;
		}
	}
	return physicalRowCounter;
}

export function reverseGapCalculator(physicalRow: number, gaps: CodeBlockGraphicData['gaps']) {
	let startLineOffset = 0;
	for (const [gapStartLine, { size }] of gaps) {
		if (physicalRow > gapStartLine + startLineOffset) {
			startLineOffset += size;
		}
	}

	return Math.max(physicalRow - startLineOffset, 0);
}

const getInstructionRegExp = (instructions: string[]) =>
	new RegExp(
		'\\b(?:' +
			instructions
				.sort((a, b) => b.length - a.length)
				.join('|')
				.replaceAll(/\*/g, '\\*')
				.replaceAll(/\]/g, '\\]')
				.replaceAll(/\[/g, '\\[') +
			')\\b',
		'd'
	);

export function generateCodeColorMap<T>(
	code: string[],
	spriteLookups: {
		fontLineNumber: T;
		fontInstruction: T;
		fontCode: T;
		fontCodeComment: T;
		fontNumbers: T;
		fontBinaryZero: T;
		fontBinaryOne: T;
	},
	instructions: string[]
): T[][] {
	return code.map(line => {
		const { index: lineNumberIndex } = /^\d+/.exec(line) || {};
		// @ts-ignore
		const { indices: instructionIndices } = getInstructionRegExp(instructions).exec(line) || { indices: [[]] };
		const { index: numberIndex } = /(?!^)(?:-|)\b(\d+|0b[01]+|0x[\dabcdef]+)\b/.exec(line) || {};
		const { index: commentIndex } = /;/.exec(line) || {};
		const binaryNumberMatch = /0b([01]+)/.exec(line) || { index: undefined };
		const { index: binaryNumberIndex } = binaryNumberMatch;
		const binaryNumber = binaryNumberMatch[1] || '';
		const binaryZeros = binaryNumber.matchAll(/(0+)/g);
		const binaryOnes = binaryNumber.matchAll(/(1+)/g);

		const codeColors = new Array(line.length).fill(undefined);

		if (typeof lineNumberIndex !== 'undefined') {
			codeColors[lineNumberIndex] = spriteLookups.fontLineNumber;
		}

		if (typeof instructionIndices !== 'undefined') {
			codeColors[instructionIndices[0][0]] = spriteLookups.fontInstruction;
			codeColors[instructionIndices[0][1]] = spriteLookups.fontCode;
		}

		if (typeof commentIndex !== 'undefined') {
			codeColors[commentIndex] = spriteLookups.fontCodeComment;
		}

		if (typeof numberIndex !== 'undefined') {
			codeColors[numberIndex] = spriteLookups.fontNumbers;
		}

		if (binaryZeros && typeof binaryNumberIndex !== 'undefined') {
			for (const match of binaryZeros) {
				codeColors[match.index + binaryNumberIndex + 2] = spriteLookups.fontBinaryZero;
			}
		}

		if (binaryOnes && typeof binaryNumberIndex !== 'undefined') {
			for (const match of binaryOnes) {
				codeColors[match.index + binaryNumberIndex + 2] = spriteLookups.fontBinaryOne;
			}
		}

		return codeColors;
	});
}
