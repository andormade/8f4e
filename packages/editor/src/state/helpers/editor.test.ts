import { backSpace, gapCalculator, moveCaret, type } from './editor';

import { ModuleGraphicData } from '../types';

const testCode = ['lorem', 'ipsum 1', 'dolor sit 0', 'amet'];

describe('moveCaret', () => {
	test('ArrowUp', () => {
		expect(moveCaret(testCode, 0, 0, 'ArrowUp')).toStrictEqual([0, 0]);
		expect(moveCaret(testCode, 1, 0, 'ArrowUp')).toStrictEqual([0, 0]);
		expect(moveCaret(testCode, 1, 7, 'ArrowUp')).toStrictEqual([0, 5]);
		expect(moveCaret(testCode, 2, 7, 'ArrowUp')).toStrictEqual([1, 7]);
	});

	test('ArrowDown', () => {
		expect(moveCaret(testCode, 0, 0, 'ArrowDown')).toStrictEqual([1, 0]);
		expect(moveCaret(testCode, 3, 0, 'ArrowDown')).toStrictEqual([3, 0]);
		expect(moveCaret(testCode, 0, 3, 'ArrowDown')).toStrictEqual([1, 3]);
		expect(moveCaret(testCode, 2, 11, 'ArrowDown')).toStrictEqual([3, 4]);
	});

	test('ArrowRight', () => {
		expect(moveCaret(testCode, 0, 0, 'ArrowRight')).toStrictEqual([0, 1]);
		expect(moveCaret(testCode, 0, 5, 'ArrowRight')).toStrictEqual([0, 5]);
	});

	test('ArrowLeft', () => {
		expect(moveCaret(testCode, 0, 0, 'ArrowLeft')).toStrictEqual([0, 0]);
		expect(moveCaret(testCode, 0, 5, 'ArrowLeft')).toStrictEqual([0, 4]);
	});
});

describe('editing', () => {
	test('backSpace', () => {
		expect(backSpace(testCode, 0, 5)).toStrictEqual({
			code: ['lore', 'ipsum 1', 'dolor sit 0', 'amet'],
			row: 0,
			col: 4,
		});

		expect(backSpace(testCode, 0, 0)).toStrictEqual({
			code: ['lorem', 'ipsum 1', 'dolor sit 0', 'amet'],
			row: 0,
			col: 0,
		});
	});

	test('type', () => {
		expect(type(testCode, 0, 0, 'a')).toStrictEqual({
			code: ['alorem', 'ipsum 1', 'dolor sit 0', 'amet'],
			row: 0,
			col: 1,
		});
	});

	test('type', () => {
		expect(type(testCode, 0, 5, 'a')).toStrictEqual({
			code: ['lorema', 'ipsum 1', 'dolor sit 0', 'amet'],
			row: 0,
			col: 6,
		});
	});
});

const gaps: ModuleGraphicData['gaps'] = new Map(
	Object.entries({ 0: { size: 2 } })
) as unknown as ModuleGraphicData['gaps'];

// describe('gapCalculator', () => {
// 	test('without gaps', () => {
// 		expect(gapCalculator(0, gaps)).toBe(2);
// 	});
// });
