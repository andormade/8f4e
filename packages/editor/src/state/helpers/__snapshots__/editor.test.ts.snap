// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateCodeColorMap  1`] = `
Array [
  Array [
    0,
    undefined,
    undefined,
    undefined,
    1,
    undefined,
    undefined,
    2,
    undefined,
    undefined,
    undefined,
    undefined,
    undefined,
    undefined,
    undefined,
    undefined,
    4,
    undefined,
  ],
]
`;
